### ISSUE TITLE
False positive reportOptionalMemberAccess when non-trivial predicate is separate from `if`

### ISSUE THREAD

**Issue by pveierland** [microsoft/pyright#7966]
**Describe the bug**
Evaluating a non-trivial predicate outside of an if-statement leads to a false positive report of member access on None. Removing the `and True` in the code below makes the example work without issue.

I'm reporting it as a bug as I'm not certain whether the shown usage is intended to be allowed, however it would be quite useful if it would work, as it would help prevent reordering certain code to accommodate the type checker.

**Code or Screenshots**

```python
from typing import Optional

def get_foo() -> Optional[str]:
    return None

foo = get_foo()

# Works:
if foo is not None and True:
    print(foo.split("_"))

# Fails:
predicate = foo is not None and True
if predicate: 
    print(foo.split("_")) # "split" is not a known attribute of "None" Pylance reportOptionalMemberAccess
```

[Pyright Playground](https://pyright-play.net/?strict=true&locale=en&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDyCMSYKAhgDYCwAUHQCYCmwUaTMA%2BsGGABQBKKAFoAfIWKkKlANoBnGCAC6ALjpQNUEBwCuIFFAByZJnTo8wUALxsO3XoLO0AxFADquANZy1tJKwtMOSgUMHxjFCYochQGKAAVEB0mX00oBBBUGD4LADo5BEokbIAiThKBASdXADFyJEofOgymBiQAY3IYKJtApGDQ8JNo2ISk0z9WFrbO7pUodU0MrJzefMLivjKKoVcSgqKYEqCQsOioT1CAdwMuxSQAIx1uqDBWEoimY7EtJhw8IgkMhUACyTAgDyYIAAgu12kw5HJqlAAAoAZXmACVwWAAG6oDAAAxicUSyUJUCuXmCqCgDzCAAsoJ05AigA)

**VS Code extension or command-line**
Running `vscode version 1.87.2 863d2581ecda6849923a2118d93a088b0745d9d6` with `Pylance v2024.5.101 (pre-release)` and a `Python 3.11.8` environment on `Linux 6.1.82 #1-NixOS SMP PREEMPT_DYNAMIC x86_64 GNU/Linux`.

**Comment by erictraut** [comment]
This behavior is by design, so I don't consider it a bug.

You're attempting to use a feature called an [aliased conditional expression](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=aliased-conditional-expression). This advanced type narrowing feature (which isn't implemented in any other Python type checkers, to my knowledge) allows you to test one variable and assign the result of that test to a second variable and then use that second variable (an "alias" for the conditional expression) later in an "if" statement to narrow the type of the first variable. While this feature works in general, it has some limitations. From the documentation:

> Unary `not` operators are allowed for expression a, but binary `and` and `or` are not.

In your code sample, you're using an `and` expression, so this expression will not work for type narrowing when used as an aliased conditional.

**Comment by pveierland** [comment]
Ok, thank you for a quick and detailed answer :+1:
